### ISSUE TITLE
Support for split archives

### ISSUE THREAD

**Issue by OndroNR** [mxmlnkn/ratarmount#36]
I could use mounting split tar archive created like this: `tar cf - ./folder | split --bytes=500MB - archive.tar.`

**Comment by mxmlnkn** [comment]
Adding this as another option to ratarmount might overload it and its command line interface. It already offers things like merging folders and such and adding a file joiner to it would just be confusing. However, you could write another FUSE script, which shows a joined view to the split file parts. Here is something I hacked together. It worked for some smaller case for me but I didn't test it extensively:

<details><summary>JoinedFileMount.py</summary>

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import bisect
import copy
import io
import os
import stat
import sys
import tempfile

import fuse

def overrides( parentClass ):
    def overrider( method ):
        assert method.__name__ in dir( parentClass )
        assert callable( getattr( parentClass, method.__name__ ) )
        return method
    return overrider

def ceilDiv( a, b ):
    return ( a + b - 1 ) // b

class JoinedFile(io.BufferedIOBase):
    """A file abstraction layer giving a joined view to one file split into parts."""

    def __init__( self, filePaths ):
        self.filePaths = []
        self.sizes = []
        self.fileobj = None
        self.currentFile = None

        for path in filePaths:
            if not os.path.isfile( path ):
                raise Exception( "File {} does not exist!".format( path ) )
            size = os.stat( path ).st_size
            if size > 0:
                self.filePaths.append( os.path.abspath( path ) )
                self.sizes.append( size )

        # Calculate cumulative sizes
        self.cumsizes = [ 0 ]
        for size in self.sizes:
            assert size > 0
            self.cumsizes.append( self.cumsizes[-1] + size )

        # Seek to the first stencil offset in the underlying file so that "read" will work out-of-the-box
        self.seek( 0 )

    def _findStencil( self, offset ):
        """
        Return index to file to which the offset belongs to. E.g., for file sizes [5,2], offsets 0 to
        and including 4 will still be inside the first file, i.e., index 0 will be returned. For offset 6,
        index 1 would be returned because it now is in the second file.
        """
        # bisect_left( value ) gives an index for a lower range: value < x for all x in list[0:i]
        # Because value >= 0 and list starts with 0 we can therefore be sure that the returned i>0
        # Consider the file sizes [2,2,2] -> cumsizes [0,2,4,6]. Seek to offset 2 should seek to the second entry.
        assert offset >= 0
        i = bisect.bisect_left( self.cumsizes, offset + 1 ) - 1
        assert i >= 0
        return i

    @overrides(io.BufferedIOBase)
    def close(self):
        self.fileobj.close()
        self.fileobj = None
        self.currentFile = None

    @overrides(io.BufferedIOBase)
    def fileno(self):
        if self.fileobj is not None:
            return self.fileobj.fileno()
        return -1

    @overrides(io.BufferedIOBase)
    def seekable(self):
        return True

    @overrides(io.BufferedIOBase)
    def readable(self):
        return True

    @overrides(io.BufferedIOBase)
    def writable(self):
        return False

    @overrides(io.BufferedIOBase)
    def read(self, size=-1):
        if size == -1:
            size = self.cumsizes[-1] - self.offset

        # This loop works in a kind of leapfrog fashion. On each even loop iteration it opens the next file
        # and on each odd iteration it reads the data and increments the offset inside the file!
        result = b''
        i = self._findStencil( self.offset )
        while size > 0 and i < len( self.sizes ):
            # Read as much as requested or as much as the current file contains
            readableSize = min( size, self.sizes[i] - ( self.offset - self.cumsizes[i] ) )
            if readableSize == 0:
                # Go to next file
                i += 1
                if i >= len( self.filePaths ):
                    break

                self.fileobj.close()
                self.fileobj = open( self.filePaths[i], 'rb' )
            else:
                # Actually read data
                tmp = self.fileobj.read( readableSize )
                self.offset += len( tmp )
                result += tmp
                size -= readableSize
                # Now, either size is 0 or readableSize will be 0 in the next iteration

        return result

    @overrides(io.BufferedIOBase)
    def seek(self, offset, whence=io.SEEK_SET):
        if whence == io.SEEK_CUR:
            self.offset += offset
        elif whence == io.SEEK_END:
            self.offset = self.cumsizes[-1] + offset
        elif whence == io.SEEK_SET:
            self.offset = offset

        if self.offset < 0:
            raise Exception("Trying to seek before the start of the file!")
        if self.offset >= self.cumsizes[-1]:
            return self.offset

        i = self._findStencil( self.offset )
        offsetInsideFile = self.offset - self.cumsizes[i]
        assert offsetInsideFile >= 0
        assert offsetInsideFile < self.sizes[i]

        # After determining where to seek to, actually do it
        if i != self.currentFile:
            if self.fileobj is not None and not self.fileobj.closed:
                self.fileobj.close()
            self.fileobj = open( self.filePaths[i], 'rb' )

        self.fileobj.seek( offsetInsideFile, io.SEEK_SET )

        return self.offset

    @overrides(io.BufferedIOBase)
    def tell(self):
        return self.offset

def testJoinedFile():
    tmpDir = tempfile.TemporaryDirectory()
    fileSizes = [2,2,2,4,8,1]
    filePaths = [ os.path.join( tmpDir.name, str(i) ) for i in range( len( fileSizes ) ) ]
    i = 0
    for path, size in zip( filePaths, fileSizes ):
        with open( path, 'wb' ) as file:
            file.write( ''.join( [ chr( i + j ) for j in range( size ) ] ).encode() )
        i += size

    print( "Test JoinedFile._findStencil" )
    joinedFile = JoinedFile( filePaths )
    expectedResults = [ 0,0, 1,1, 2,2, 3,3,3,3, 4,4,4,4,4,4,4,4, 5 ]
    for offset, iExpectedStencil in enumerate( expectedResults ):
        assert joinedFile._findStencil( offset ) == iExpectedStencil

    print( "Test JoinedFile with single file" )

    assert JoinedFile( [ filePaths[0] ] ).read( 1 ) == b"\x00"
    assert JoinedFile( [ filePaths[0] ] ).read( 2 ) == b"\x00\x01"
    assert JoinedFile( [ filePaths[0] ] ).read() == b"\x00\x01"

    print( "Test JoinedFile using two files" )

    joinedFile = JoinedFile( filePaths[:2] )
    assert joinedFile.read() == b"\x00\x01\x02\x03"
    for i in [0,1,2,3,2,1,0,2,0,2]:
        joinedFile.seek( i )
        joinedFile.tell() == i
        assert joinedFile.read( 1 ) == chr( i ).encode()
    joinedFile.seek( 0, io.SEEK_END )
    assert joinedFile.tell() == 4


class FuseFileObjectMount( fuse.Operations ):
    def __init__( self, fileobj, mountPoint, stats ):
        self.fileobj = fileobj
        self.stats = stats
        self.fileName = 'joined'

        # Get file size
        oldPos = self.fileobj.tell()
        self.fileobj.seek( 0, io.SEEK_END )
        self.stats['st_size'] = fileobj.tell()
        self.stats['st_blocks'] = ceilDiv( self.stats['st_size'], self.stats['st_blksize'] )
        del self.stats['st_ino']
        self.fileobj.seek( oldPos )

        # Create mount point if it does not exist
        self.mountPointWasCreated = False
        if mountPoint and not os.path.exists( mountPoint ):
            os.mkdir( mountPoint )
            self.mountPointWasCreated = True
        self.mountPoint = os.path.realpath( mountPoint )

    def __del__( self ):
        try:
            if self.mountPointWasCreated:
                os.rmdir( self.mountPoint )
        except:
            pass

    @overrides( fuse.Operations )
    def getattr( self, path, fh = None ):
        if path == '/':
            stats = copy.deepcopy( self.stats )
            stats['st_mode'] = 0o777 | stat.S_IFDIR
            stats['st_size'] = 0
            return stats

        if path == '/' + self.fileName:
            return self.stats

        raise fuse.FuseOSError( fuse.errno.ENOENT )

    @overrides( fuse.Operations )
    def readdir( self, path, fh ):
        if path != '/':
            raise fuse.FuseOSError( fuse.errno.ENOENT )
        return [ '.', '..', self.fileName ]

    @overrides( fuse.Operations )
    def readlink( self, path ):
        return ""

    @overrides( fuse.Operations )
    def read( self, path, length, offset, fh ):
        if path != '/' + self.fileName:
            raise fuse.FuseOSError( fuse.errno.ENOENT )

        try:
            self.fileobj.seek( offset, os.SEEK_SET )
            return self.fileobj.read( length )
        except RuntimeError as e:
            traceback.print_exc()
            print( "Caught exception when trying to read data from underlying TAR file! Returning errno.EIO." )
            raise fuse.FuseOSError( fuse.errno.EIO )

def cli( args = None ):
    if not args:
        testJoinedFile()
        return

    assert len( args ) >= 2
    mountPoint = args[-1]
    joinedFile = JoinedFile( args[:-1] )

    stats = os.stat( args[0] )
    stats = dict( [ field, getattr( stats, field ) ] for field in dir( stats ) if field.startswith( 'st_' ) )

    fuseOperationsObject = FuseFileObjectMount( joinedFile, mountPoint, stats )

    print( "Join", args[:-1], "into", os.path.join( mountPoint + '/' + fuseOperationsObject.fileName ) )

    fuse.FUSE( operations = fuseOperationsObject,
               mountpoint = mountPoint,
               foreground = False,
               nothreads  = True,
               debug      = False )

if __name__ == '__main__':
    cli( sys.argv[1:] )
```

</details>

This is how you would use it:

```bash
python3 JoinedFileMount.py
    Test JoinedFile._findStencil
    Test JoinedFile with single file
    Test JoinedFile using two files
head -c 10000 /dev/urandom > largeFile
split --bytes=3000 largeFile part
python3 JoinedFileMount.py part* tmpmountpoint
    Join ['partaa', 'partab', 'partac', 'partad'] into tmpmountpoint/joined
diff --report-identical-files largeFile tmpmountpoint/joined
    Files largeFile and tmpmountpoint/joined are identical
ratarmount tmpmountpoint/joined mountedTar
fusermount -u tmpmountpoint
```

I know this is troublesome but I don't see how I can integrate this functionality into ratarmount for now. It might be best to start a new package on PyPI for this functionality.

**Comment by OndroNR** [comment]
Thank you, I will try this script. Making 'union view' as separate tool makes sense. It could have more general use.

**Comment by mxmlnkn** [comment]
There also are [concatfs](https://github.com/schlaile/concatfs) and [merged-fuse](https://github.com/fly-studio/merged-fuse), which basically do the same as my python script (which can also be found in the [joinFuseLayer branch](https://github.com/mxmlnkn/ratarmount/tree/joinFuseLayer). Someone requested this again today, which got me thinking again whether to include this in ratarmount or how to add this to the already complex CLI.

**Comment by JuniorJPDJ** [comment]
This was me!

If you want to add this to ratarmount you could just add `--detect-split-archives` and lookup all files with the same name but other extensions and then sort them in order.
Would be also cool to provide some sort of regexp for name of merging files.
Eg. in my case split files are named `\d*_*.\.zip\..*`.

**Comment by mxmlnkn** [comment]
I implemented a prototype on the develop branch. You can try it out with:

```bash
python3 -m pip install --user --force-reinstall 'git+https://github.com/mxmlnkn/ratarmount.git@develop#egginfo=ratarmountcore&subdirectory=core'
python3 -m pip install --user 'git+https://github.com/mxmlnkn/ratarmount.git@develop#egginfo=ratarmount'
```

It can detect decimal, hexadecimal, and alphabetical split extensions of arbitrary length. But it assumes that the split extension sequence either begins at 0 or 1 even though `split` theoretically allows setting `FROM` higher than 1. Furthermore, by default `split` will create files without any dot but ratarmount will only look for extensions, i.e., the prefix given to `split` is required to end with `.` for compatibility.

It does not yet work recursively if there is a split archive inside another archive but I'll probably implement that before a release.

I also decided to make the split file detection the default behavior and offer a `--no-detect-split` option. This option is also not yet implemented.

You can try it out with:

```bash
base64 /dev/urandom | head -c $(( 64 * 1024 )) > random
tar -cf- random | split --bytes=32K --numeric-suffixes=1 --suffix-length=3 - random.tar.
ls -la
# random.tar.001
# random.tar.002
# random.tar.003
ratarmount random.tar.001 mountPoint
diff -s random mountPoint/random
# Files random and mountPoint/random are identical
```

In this example, I specified the `.001` file but it also works with any other part, e.g., `.002`. It even should work when specifying all of the parts `ratarmount random.tar.[0-9][0-9][0-9] mountPoint`. But the parts should not be interspersed with other archives to be union mounted because the order is important for the resulting merged view. E.g.: `ratarmount foo.tar random.tar.001 bar.tar random.tar.002 mountPoint`. In that case, ratarmount will print an error because the "same" `random.tar` archive appears at conflicting positions.

**Comment by mxmlnkn** [comment]
Closed with d5bb6e4 in master. The release should follow in the next weeks.
